### PR TITLE
Create function variables using the data produced by the pre-scanner.

### PR DESCRIPTION
**Pull request by zherczeg**
The pre-scanner now is able to track all variable declarations and produces a compressed stream which store this data for each function and catch block. When a function or catch block is parsed, this information is decoded and the appropriate variables are created. Furthermore a stack scope is created which contains the currently available local variables and their register or literal index.

### CODE REVIEW

**Comment by renatahodovan** [comment]
###### JerryScript revision
7cf5edf3


###### Build platform
Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
var $ = { $() { } } = { g ( ) { m => m ++ ( ) = { constructor ( 
```


###### Output
```text
ICE: Assertion 'context_p->next_scanner_info_p->type == SCANNER_TYPE_FUNCTION' failed at jerryscript/jerry-core/parser/js/js-parser.c(parser_parse_function_arguments):2100.
Error: ERR_FAILED_INTERNAL_ASSERTION
```


###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x56579a6a in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x56612442 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x56612483 in jerry_assert_fail (assertion=0x566bcee0 "context_p->next_scanner_info_p->type == SCANNER_TYPE_FUNCTION", file=0x566bc0e0 "jerryscript/jerry-core/parser/js/js-parser.c", function=0x56695120 <__func__.5501.lto_priv.464> "parser_parse_function_arguments", line=2100) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x5661f8bb in parser_parse_function_arguments (context_p=0xffffc4b0, end_type=LEXER_RIGHT_PAREN) at jerryscript/jerry-core/parser/js/js-parser.c:2100
#7  0x566235f9 in parser_parse_function (context_p=0xffffc4b0, status_flags=3146758) at jerryscript/jerry-core/parser/js/js-parser.c:2623
#8  0x5659c24d in parser_parse_class_literal (context_p=0xffffc4b0) at jerryscript/jerry-core/parser/js/js-parser-expr.c:514
#9  0x5659ccc6 in parser_parse_class (context_p=0xffffc4b0, is_statement=false) at jerryscript/jerry-core/parser/js/js-parser-expr.c:649
#10 0x5659fe49 in parser_parse_unary_expression (context_p=0xffffc4b0, grouping_level_p=0xffffc050) at jerryscript/jerry-core/parser/js/js-parser-expr.c:1376
#11 0x565a4230 in parser_parse_expression (context_p=0xffffc4b0, options=1) at jerryscript/jerry-core/parser/js/js-parser-expr.c:2149
#12 0x565a3bf4 in parser_parse_block_expression (context_p=0xffffc4b0, options=0) at jerryscript/jerry-core/parser/js/js-parser-expr.c:2091
#13 0x565707e3 in parser_parse_statements (context_p=0xffffc4b0) at jerryscript/jerry-core/parser/js/js-parser-statm.c:2547
#14 0x5662188e in parser_parse_source (arg_list_p=0x0, arg_list_size=0, source_p=0x566f9fc0 <buffer.lto_priv> "var id_0 ; \nvar called = false ; \nfunction f ( ) { } \nvar B = class extends f { constructor ( ) { eval ( ) ; \nsuper ( id_1 ) \nthis .g ( ) \n} } \nC = class extends B { g ( ) { ( m ) => m ++ ( ) \n} } \nD "..., source_size=319, parse_opts=0, error_location_p=0xffffc6c0) at jerryscript/jerry-core/parser/js/js-parser.c:2407
#15 0x566241ae in parser_parse_script (arg_list_p=0x0, arg_list_size=0, source_p=0x566f9fc0 <buffer.lto_priv> "var id_0 ; \nvar called = false ; \nfunction f ( ) { } \nvar B = class extends f { constructor ( ) { eval ( ) ; \nsuper ( id_1 ) \nthis .g ( ) \n} } \nC = class extends B { g ( ) { ( m ) => m ++ ( ) \n} } \nD "..., source_size=319, parse_opts=0, bytecode_data_p=0xffffc780) at jerryscript/jerry-core/parser/js/js-parser.c:2848
#16 0x56670e34 in jerry_parse (resource_name_p=0xffffcd8e "/home/reni/.fuzzinator_35486//jerryscript/picireny/33680886534313296745214480367664733813.js", resource_name_length=92, source_p=0x566f9fc0 <buffer.lto_priv> "var id_0 ; \nvar called = false ; \nfunction f ( ) { } \nvar B = class extends f { constructor ( ) { eval ( ) ; \nsuper ( id_1 ) \nthis .g ( ) \n} } \nC = class extends B { g ( ) { ( m ) => m ++ ( ) \n} } \nD "..., source_size=319, parse_opts=0) at jerryscript/jerry-core/api/jerry.c:426
#17 0x5666dc2d in main (argc=2, argv=0xffffcb14) at jerryscript/jerry-main/main-unix.c:726
```


<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

File: jerry-core/parser/js/js-scanner-util.c [hunk 1]
```diff
@@ -225,6 +308,548 @@ scanner_seek (parser_context_t *context_p) /**< context */
...
+  *literal_p = context_p->token.lit_location;
+
+  literal_p->type = SCANNER_LITERAL_IS_ARG;
+} /* scanner_append_argument */
+
+/**
+ * Check whether the last identifier is an "eval".
+ */
+void
+scanner_check_eval (parser_context_t *context_p, /**< context */
```
Comment (galpeter) [inline jerry-core/parser/js/js-scanner-util.c]: IMHO the name is misleading: If it checks something then it should return something. This method checks the 'eval' literal but it also sets a status flag.

File: jerry-core/parser/js/js-scanner.c [hunk 2]
```diff
@@ -197,6 +208,64 @@ scanner_process_simple_arrow (parser_context_t *context_p, /**< context */
   scanner_check_arrow_body (context_p, scanner_context_p);
 } /* scanner_process_simple_arrow */
 
+/**
+ * Process arrow function with a single argument.
```
Comment (galpeter) [inline jerry-core/parser/js/js-scanner.c]: Arrow function with one argument? What happens if there are more arguments?

File: jerry-core/parser/js/js-scanner.c [hunk 3]
```diff
@@ -2065,6 +2352,103 @@ scanner_scan_all (parser_context_t *context_p, /**< context */
...
         }
+        case SCANNER_TYPE_FUNCTION:
+        case SCANNER_TYPE_BLOCK:
+        {
+          const uint8_t *prev_source_p = info_p->source_p - 1;
+          const uint8_t *data_p;
+
+          if (info_p->type == SCANNER_TYPE_FUNCTION)
+          {
+
```
Comment (galpeter) [inline jerry-core/parser/js/js-scanner.c]: Do we need the extra newline here?

File: tests/unit-core/test-snapshot.c [hunk 4]
```diff
@@ -226,7 +226,7 @@ main (void)
       0x4A, 0x52, 0x52, 0x59, 0x18, 0x00, 0x00, 0x00,
       0x00, 0x00, 0x00, 0x00, 0x48, 0x00, 0x00, 0x00,
       0x01, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00,
-      0x03, 0x00, 0x01, 0x00, 0x01, 0x00, 0x01, 0x00,
+      0x03, 0x00, 0x01, 0x00, 0x21, 0x00, 0x01, 0x00,
```
Comment (galpeter) [inline tests/unit-core/test-snapshot.c]: If the snapshot changed, shouldn't the snapshot version also change?

Comment (zherczeg) [reply tests/unit-core/test-snapshot.c]: It is done.

File: jerry-core/parser/js/js-scanner-util.c (hunk 1 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner-util.c]: What is your suggestion? As for me `is_eval` returns with something, but anything is good for me.

File: jerry-core/parser/js/js-scanner.c (hunk 2 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner.c]: Wrong comment. Copied from elsewhere.

File: jerry-core/parser/js/js-scanner.c (hunk 3 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner.c]: Removed

**Comment by zherczeg** [comment]
@renatahodovan your issue is unrelated, #3215 should fix it.

File: jerry-core/parser/js/js-scanner.h [hunk 5]
```diff
@@ -65,6 +64,8 @@ typedef struct scanner_info_t
   struct scanner_info_t *next_p; /**< next info structure */
   const uint8_t *source_p; /**< triggering position of this scanner info */
   uint8_t type; /**< type of the scanner info */
+  uint16_t u8_arg; /**< custom 8-bite value */
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner.h]: typo, `bite`

File: jerry-core/parser/js/js-scanner-internal.h [hunk 6]
```diff
@@ -16,7 +16,7 @@
 #ifndef JS_SCANNER_INTERNAL_H
 #define JS_SCANNER_INTERNAL_H
 
-/** \addtogroup parser Parser
+/* \addtogroup parser Parser
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-internal.h]: Is this change really needed here?

File: jerry-core/parser/js/js-parser-util.c [hunk 7]
```diff
@@ -85,6 +85,51 @@ parser_emit_two_bytes (parser_context_t *context_p, /**< context */
...
   (context_p)->byte_code.last_p->bytes[(context_p)->byte_code.last_position++] = (uint8_t) (byte)
 
+#if ENABLED (JERRY_PARSER_DUMP_BYTE_CODE)
+
+/**
+ * Print literal corresponding to the current index
+ */
+static void
+parser_print_literal (parser_context_t *context_p, /**< context */
+                      uint16_t literal_index)
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-util.c]: Missing argument description.

File: jerry-core/parser/js/js-scanner-util.c (hunk 1 above)
Comment (galpeter) [reply jerry-core/parser/js/js-scanner-util.c]: What if `scanner_update_eval` ? Also the function's comment needs update, to detail that it'll check the eval identifier and update the flag.

File: jerry-core/parser/js/js-scanner-internal.h (hunk 6 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner-internal.h]: Somehow I get doxygen error messages without it.

File: jerry-core/parser/js/js-parser-internal.h [hunk 8]
```diff
@@ -365,8 +390,13 @@ typedef struct
 {
   PARSER_TRY_CONTEXT (try_buffer);            /**< try_buffer */
   parser_error_t error;                       /**< error code */
-  void *allocated_buffer_p;                   /**< dinamically allocated buffer
+  /** Union for rarely used members. */
+  union
+  {
+    void *allocated_buffer_p;                 /**< dinamically allocated buffer
                                                *   which needs to be freed on error */
+    scanner_context_t *scanner_context_p;     /**< scanner fontext for the pre-scanner */
```
Comment (dbatyai) [inline jerry-core/parser/js/js-parser-internal.h]: Typo: fontext

File: jerry-core/parser/js/js-scanner-util.c (hunk 1 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner-util.c]: I changed it to `scanner_detect_eval_call`, what do you think?

File: jerry-core/parser/js/js-scanner.h (hunk 5 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner.h]: (And it is u16 instead of u8)

File: jerry-core/vm/vm.c [hunk 9]
```diff
@@ -797,90 +797,70 @@ vm_init_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
 
-        while (literal_index <= literal_index_end)
-        {
-          ecma_string_t *name_p = ecma_get_string_from_value (literal_start_p[literal_index]);
-          vm_var_decl (frame_ctx_p, name_p);
-          literal_index++;
-        }
+        ecma_string_t *name_p = ecma_get_string_from_value (literal_start_p[literal_index]);
+        vm_var_decl (frame_ctx_p, name_p);
+        literal_index++;
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply jerry-core/vm/vm.c]: Old code :)
